### PR TITLE
Feature/scatterplot heatmap

### PR DESCRIPTION
**Pull request by adrianmroz-allegro**

### CODE REVIEW

File: src/client/visualizations/scatterplot/heatmap.tsx [hunk 1]
```diff
@@ -42,56 +43,89 @@ interface HeatmapProps {
...
+  xScale,
+  yScale,
+  fillColor
+}) => {
+  const [x0, x1] = xRange;
+  const [y0, y1] = yRange;
+  const xPosition = xScale(x0);
+  const width = xScale(x1) - xPosition;
+  const yPosition = yScale(y1);
+  const height = yScale(y0) - yScale(y1);
```
Comment (adrianmroz-allegro) [inline src/client/visualizations/scatterplot/heatmap.tsx:122]:
```suggestion
  const height = yScale(y0) - yPosition;
```

**Comment by kzadurska** [comment]
Zrzut ekranu 2022-02-23 o 13 49 05
Found [this case](http://localhost:9090/#2019-weather/4/N4IgbglgzgrghgGwgLzgFwgewHYgFwhQDG6aApgE4AOCmaIANONPEqhjgMplobYDmUfKCgALTAHdOMALYy4FAJ740FGGQC+TDDLLIcZfCACiaIgHoAqgBUAwoxAAzCAnIUheANqg0iqoYIdQyYKMkcjIIcggAVKLAATD28QeJgKdCxcAmiARgBNByhyKnwAWhztPwCQBHQyIpANAF1mpigaCDQknyqjIooIAQd4iF1sKEyjIhwRjnHCzAp6PFBQ8IIZQYB9chkSyv8+uPrhiFCiOaN4+qIybBGhpn8BzHijRqYkTeXsGAQEVqEY7dEBrSh3W5GTbYHZkPYORyLeTLHqHAjXRxwP70JhgRDqd5aEC+NEgXRwWChD6rMLg7CQghUc4QKidDI4BFI9DCYm9dFhLGuBx4hAEggfXmk8mUwyA1nYbBkeIAEVGdwmOCSTSeg0V8U4i2WZO2uxKGiAA), where our calculations are not correct even without changing the screen size 😳

File: src/client/visualizations/scatterplot/scatterplot.tsx [hunk 2]
```diff
@@ -103,27 +103,31 @@ export class Scatterplot extends React.Component<ChartProps, ScatterplotState> {
...
+        <XAxis
+          scale={xScale}
+          stage={calculateXAxisStage(plottingStage)}
+          ticks={getTicksByDimension(xTicks, plottingStage.width)}
+          formatter={xSeries.formatter()}
+          tickSize={TICK_SIZE}/>
         <YAxis
           stage={calculateYAxisStage(plottingStage)}
-          ticks={yTicks}
+          ticks={getTicksByDimension(yTicks, plottingStage.height)}
```
Comment (adrianmroz-allegro) [inline src/client/visualizations/scatterplot/scatterplot.tsx:114]: I don't like this name - we have very specific thing in our domain language that is called dimension. `getTicksForAvailableSpace`?

Comment (kzadurska) [reply src/client/visualizations/scatterplot/scatterplot.tsx]: Sure, that's much better, done ✅

**Review by mkuthan — COMMENTED** [review]
A few questions:
1) blue colour for the summary - intentional or not?
2) settings label is "Show heatmap summary" but legend label is just a "count" - could we make it more consistent?
3) have you checked CSS theming compatibility?
4) have you checked "lower is better" for timeshift?

**Comment by kzadurska** [comment]
Note to self, another issue to look into:
Heatmap vs pinboard [[klik](http://localhost:9090/#covid19/4/N4IgbglgzgrghgGwgLzgFwgewHYgFwhQDG6aApgE4AOCmaIANONPEqhjgMplobYDmUfKCgALTAHdOMALYy4FAJ740FGGQC+TDDLLIcZfCACiaIgHoAqgBUAwoxAAzCAnIUheANqg0iqoYIdQyYKMkcjIIcggAVKLAATD28QeJgKdCxcAmiARgBNByhyKnwAWhztPwCQBHQyIpANAF0tHyqjIooIAQdQ8IJaEg5cJjgiYaNuhzBEdSSQS2wIcniAAgBpbv54zBkQJqZsOnxHRChNA8IaZaS2/w7VLYd4iF1sKEyjQYycQswKeh4UB9Iwkc5CSr3AjnLr1Z4QULjT4EeL1IhkbAvHpMfxdTDxIyNJhIGTLfDYGAIBDNJgwiBwrzAsKUDHo0FwcEORz/eSAu7VVGnSn0JgzBDqQlaEC+KEgXQctKGVogPos7BslFkdCiCFOHnoYTS9qaoWuaazapEo2y+WwUKNS5UbrYMjxAAirwxHxw82+E0dztdnH+gJAYLhGiAA=)]


https://user-images.githubusercontent.com/22297591/155513696-87ec2482-7949-41cf-94ac-0f2a8eea5c03.mov

**Comment by kzadurska** [comment]
@mkuthan 

> 1. blue colour for the summary - intentional or not?
> 3. have you checked CSS theming compatibility?

Blue is intentional:
- contrast to orange
- similar in hue to item-measure tiles above

Currently Turnilo does not have support for a secondary color. And I think we don't have a possibility to pass to TS colors defined in CSS. I'd leave it as is for now. Created na issue https://github.com/allegro/turnilo/issues/852 so that it is addressed.

> 2. settings label is "Show heatmap summary" but legend label is just a "count" - could we make it more consistent?

Renamed to "Count per summary bin" so that it includes _summary_, but is it enough or should we use split title here in some form as well? Naming is the hardest part, we can brainstorm this, if you want. I'm open to suggestions :) 

> 4. have you checked "lower is better" for timeshift?

In https://github.com/allegro/turnilo/pull/843/files I used `SeriesBubbleContent`, which in turn uses `MeasureBubbleContent`, where "lower is better" setting is passed, take a look here https://github.com/allegro/turnilo/blob/master/src/client/components/series-bubble-content/series-bubble-content.tsx#L39

File: src/client/visualizations/scatterplot/utils/get-plotting-data.ts [hunk 3]
```diff
@@ -56,7 +57,21 @@ export function preparePlottingData(data: Dataset, essence: Essence, stage: Stag
...
+}
+
+function extendExtentIfNeeded(extent: number[]): number[] {
+  const [rangeStart, rangeEnd] = extent;
+
+  if (rangeStart !== rangeEnd) {
+    return extent;
+  }
+
+  const loweredRangeStart = Math.floor(rangeStart - rangeStart * EXTENT_EXTEND_FACTOR);
```
Comment (adrianmroz-allegro) [inline src/client/visualizations/scatterplot/utils/get-plotting-data.ts:72]: Isn't `nice` enough?

Comment (kzadurska) [reply src/client/visualizations/scatterplot/utils/get-plotting-data.ts]: It was, fixed.

**Pull request merged**